doc/proposals/scorecard-user-tests.md

```diff
@@ -0,0 +1,112 @@
+
+[Goals](#Goals)
+
+[Design overview](#Design_overview)
+
+[User facing usage](#User_facing_usage)
+
+[Observations and open questions](#Observations_and_open_questions)
+
+## Background
+
+The operator scorecard is intended to allow users to run a generic set of tests on their operators. Some simple checks can be performed, but more complicated
+tests verifying that the operator actually works are not possible to do in a generic way. This leads to some of the functional tests in the current scorecard
+implementation to be inaccurate or too insufficient to be useful. For more useful functional scorecard tests, we need to allow some basic user input for tests.
+
+## Goals
+
+- Implement user-defined scorecard tests
+- Replace existing "Operator actions are reflected in status" and "Writing into CRs has effects" tests with user-defined tests
+
+## Design overview
+
+A new section would be added to the scorecard config file called `functional_tests` where a user can define various aspects of a test. For example, this definition runs a similar test to the memcached-operator scale test from the SDK's e2e test:
+
+```yaml
+functional_tests:
+  - cr: "deploy/crds/memcached.cr.yaml"
+    expected_resources:
+      - kind: Deployment
+        name: "example_memcached"
+        fields:
+          status:
+            readyReplicas: 3
+          spec:
+            template:
+              spec:
+                containers:
+                  - image: memcached:1.4.36-alpine
+    expected_status:
+      scorecard_function_length:
+        nodes: 3
+    modifications:
+      - spec:
+        - size: 4
+        expected_resources:
+          - kind: Deployment
+            name: "example_memcached"
+            fields:
+              status:
+                readyReplicas: 4
+        expected_status:
+          scorecard_function_length:
+            nodes: 4
+
+```
+
+This is what the golang structs would look like, including comments describing each field:
+
+```go
+// Struct containing a user defined test. User passes tests as an array using the `functional_tests` viper config
+type UserDefinedTest struct {
+    // Path to cr to be used for testing
+    CRPath string `mapstructure:"cr"`
+    // Resources expected to be created after the operator reacts to the CR
+    ExpectedResources []ExpectedResource `mapstructure:"expected_resources"`
+    // Expected values in CR's status after the operator reacts to the CR
+    ExpectedStatus map[string]interface{} `mapstructure:"expected_status"`
+    // Sub-tests modifying a few fields with expected changes
+    Modifications []Modification `mapstructure:"modifications"`
+}
+
+// Struct containing a resource and its expected fields
+type ExpectedResource struct {
+    // Kind of resource
+    Kind string `mapstructure:"kind"`
+    // Name of resource
+    Name string `mapstructure:"name"`
+    // The fields we expect to see in this resource
+    Fields map[string]interface{} `mapstructure:"fields"`
+}
```

Comment:
There are probably complexities I'm not thinking of, but I wonder if we could define `ExpectedResource` like one of these:

```go
type ExpectedResource unstructured.Unstructured
type ExpectedResource map[string]interface{}
```

Being able to mostly copy/paste the output of `kubectl get mykind myresource -o yaml` into the test case seems more ergonomic and predictable.

Reply:
The main issue there is that it might cause problems with the `scorecard_function_` fields. I'll try to see if it works though, because it would be nice to be able to simplify the spec.

Reply:
I did some testing and was able to get is working with `Expected.Resources` being `[]map[string]interface{}`, which simplifies the implementation a bit. I've updated the config to reflect that.